conceptual/Npgsql/diagnostics/tracing.md

```diff
@@ -29,3 +30,30 @@ For example, Zipkin visualizes traces in the following way:
 ![Zipkin UI Sample](/img/zipkin.png)
 
 In this trace, the Npgsql query (to database testdb) took around 800ms, and was nested inside the application's `work1` activity, which also had another unrelated `subtask1`. This allows understanding the relationships between the different activities, and where time is being spent.
+
+## Configuration options
+
+> [!NOTE]
+>
+> This feature was introduced in Npgsql 9.0
+
+Once you've enabled Npgsql tracing as above, you can tweak its configuration via the <xref:Npgsql.NpgsqlDataSourceBuilder.ConfigureTracingOptions*?displayProperty=nameWithType> API:
+
+```csharp
+dataSourceBuilder.ConfigureTracingOptions(new NpgsqlTracingOptions
+{
+    // Set the command SQL as the span name
+    ProvideSpanNameForCommand = cmd => cmd.CommandText,
+    // Filter out COMMIT commands
+    FilterCommand = cmd => !cmd.CommandText.StartsWith("COMMIT", StringComparison.Ordinal)
+});
+```
+
+This allows you to:
+
+* Specify a filter which determines which commands get traced
+* Set the the tracing span name (e.g. use the command's SQL as the span name)
+* Add arbitrary tags to the tracing span, based on the command
+* Disable the time-to-first-read event that's emitted in spans
+
+.NET's [`AsyncLocal`](https://learn.microsoft.com/dotnet/api/system.threading.asynclocal-1) allows you to flow arbitrary information from the command call-site (where you create and execute the command) to the above callbacks; for example, you can give each command a unique logical name which would be set as the span name. Note that separate callbacks need to be registered for <xref:Npgsql.NpgsqlCommand> and <xref:Npgsql.NpgsqlBatch>.
```

Comment:
Do you think we should add an example on how to use it? I'm mostly worried about sync queries, where you have to clear after a query.

Reply:
Yeah, you're right that this merits a code sample - I'll add that. But can you provide more details on what you're worried about with sync queries and clearing?

BTW for sync queries BTW I'd expect users to use thread-local storage (e.g. [ThreadLocal](https://learn.microsoft.com/en-us/dotnet/api/system.threading.threadlocal-1)) rather than AsyncLocal (we can show this in a code sample too).

/cc @NinoFloris

Reply:
> But can you provide more details on what you're worried about with sync queries and clearing?

Mostly how once you set it, you have to clear that async local value, as otherwise it's going to be stuck (unless you're in async method, where the runtime will do that automatically).

> BTW for sync queries BTW I'd expect users to use thread-local storage (e.g. [ThreadLocal](https://learn.microsoft.com/en-us/dotnet/api/system.threading.threadlocal-1)) rather than AsyncLocal (we can show this in a code sample too).

Probably, yeah, though it can suffer from the same issue (once it's set, it's not going to be cleared automatically, leaking the settings you're trying to pass).

Reply:
Shouldn't we mostly want to point people to Activity.Baggage? 

Regarding AsyncLocals: they replaced the .NET Framework CallContext concept, so in that sense they're not strictly *async only* (you could call them DispatchableLocals too) Both approaches store their values on the ExecutionContext that's captured and flown along dispatches; there's mostly a difference in how the locals are tracked. In all cases when executed/resumed the context will be restored onto a thread static.

A strong argument for using one mechanism over one for async and another for sync is that users often mix async and sync. Even if they think they don't libraries like EF and dapper might still do sync calls at times. Having to keep both in sync is likely unexpected and error prone.

Reply:
>Mostly how once you set it, you have to clear that async local value, as otherwise it's going to be stuck (unless you're in async method, where the runtime will do that automatically).

Yeah I think AsyncLocals have an advantage here too, every TP work item gets a clean ExecutionContext, so they'll get cleared at some point. ThreadStatics are entirely untracked though.